src/rotativeperformancelogger.cpp

```diff
@@ -0,0 +1,812 @@
+            {
+                outputFile.close();
+                return;  // This request means we have received a termination signal; close and exit the thread as quick & clean as possible
+            }
+        }
+    }
+
+};
+
+
+RotativePerformanceLogger::RotativePerformanceLogger()
+{
+}
+
+RotativePerformanceLogger::~RotativePerformanceLogger()
+{
+    MegaApi::removeLoggerObject(this); // after this no more calls to RotativePerformanceLogger::log
+    {
+        std::lock_guard<std::mutex> g(mLoggingThread->mLogMutex);
+        mLoggingThread->mLogExit = true;
+        mLoggingThread->mLogConditionVariable.notify_one();
+    }
+    mLoggingThread->mLogThread->join();
+    mLoggingThread->mLogThread.reset();
+}
+
+void RotativePerformanceLogger::initialize(const char * logsPath, const char * logFileName, bool logToStdout)
+{
+    auto logsPathLocalPath = LocalPath::fromPlatformEncoded(logsPath);
+    auto logFileNameLocalPath = LocalPath::fromPlatformEncoded(logFileName);
+
+    mLogToStdout = logToStdout;
+
+    MegaFileSystemAccess *fsAccess = new MegaFileSystemAccess();
+    fsAccess->mkdirlocal(logsPathLocalPath, false);
+
+    mLoggingThread.reset(new RotativePerformanceLoggerLoggingThread());
+    mLoggingThread->startLoggingThread(logsPathLocalPath, logFileNameLocalPath);
+
+    MegaApi::setLogLevel(MegaApi::LOG_LEVEL_MAX);
+    MegaApi::addLoggerObject(this);
+}
+
+RotativePerformanceLogger& RotativePerformanceLogger::Instance() {
+    static RotativePerformanceLogger myInstance;
+    return myInstance;
+}
+
+void RotativePerformanceLogger::setArchiveNumbered()
+{
+    mLoggingThread->mArchiveType = archiveTypeNumbered;
+}
+
+void RotativePerformanceLogger::setArchiveTimestamps(long int maxFileAgeSeconds)
+{
+    mLoggingThread->mArchiveType = archiveTypeTimestamp;
+    mLoggingThread->archiveMaxFileAgeSeconds = maxFileAgeSeconds;
+}
+
+inline void twodigit(char*& s, int n)
+{
+    *s++ = static_cast<char>(n / 10 + '0');
+    *s++ = static_cast<char>(n % 10 + '0');
+}
+
+char* filltime(char* s, struct tm*  gmt, int microsec)
+{
+    twodigit(s, gmt->tm_mday);
+    *s++ = '/';
+    twodigit(s, gmt->tm_mon + 1);
+    *s++ = '/';
+    twodigit(s, gmt->tm_year % 100);
+    *s++ = '-';
+    twodigit(s, gmt->tm_hour);
+    *s++ = ':';
+    twodigit(s, gmt->tm_min);
+    *s++ = ':';
+    twodigit(s, gmt->tm_sec);
+    *s++ = '.';
+    s[5] = static_cast<char>(microsec % 10 + '0');
+    s[4] = static_cast<char>((microsec /= 10) % 10 + '0');
+    s[3] = static_cast<char>((microsec /= 10) % 10 + '0');
+    s[2] = static_cast<char>((microsec /= 10) % 10 + '0');
+    s[1] = static_cast<char>((microsec /= 10) % 10 + '0');
+    s[0] = static_cast<char>((microsec /= 10) % 10 + '0');
+    s += 6;
+    *s++ = ' ';
+    *s = 0;
+    return s;
+}
+
+std::mutex threadNameMutex;
+std::map<std::thread::id, std::string> threadNames;
+struct tm lastTm;
+time_t lastT = 0;
+std::thread::id lastThreadId;
+const char* lastThreadName;
```

Comment:
we probably want to move this code and helper functions to a class

Reply:
Moved to a new helper class.